File: java/iroha_android/src/main/java/org/hyperledger/iroha/android/sorafs/GatewayFetchOptions.java
package org.hyperledger.iroha.android.sorafs;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Objects;

/**
 * Helper for composing SoraFS gateway fetch options.
 *
 * <p>The builder exposes the subset of orchestrator configuration that Android clients commonly
 * tweak (telemetry labels, provider limits, retry budgets, transport/anonymity policy). The helper
 * serialises to a map matching the Norito JSON structure consumed by the Rust orchestrator.
 */
public final class GatewayFetchOptions {
  private final String manifestEnvelopeBase64;
  private final String manifestCidHex;
  private final String clientId;
  private final String telemetryRegion;
  private final String rolloutPhase;
  private final Integer maxPeers;
  private final Integer retryBudget;
  private final TransportPolicy transportPolicy;
  private final AnonymityPolicy anonymityPolicy;
  private final WriteModeHint writeModeHint;

  private GatewayFetchOptions(final Builder builder) {
    this.manifestEnvelopeBase64 = builder.manifestEnvelopeBase64;
    this.manifestCidHex = builder.manifestCidHex;
    this.clientId = builder.clientId;
    this.telemetryRegion = builder.telemetryRegion;
    this.rolloutPhase = builder.rolloutPhase;
    this.maxPeers = builder.maxPeers;
    this.retryBudget = builder.retryBudget;
    this.transportPolicy = builder.transportPolicy;
    this.anonymityPolicy = builder.anonymityPolicy;
    this.writeModeHint = builder.writeModeHint;
  }

  public String manifestEnvelopeBase64() {
    return manifestEnvelopeBase64;
  }

  public String manifestCidHex() {
    return manifestCidHex;
  }

  public String clientId() {
    return clientId;
  }

  public String telemetryRegion() {
    return telemetryRegion;
  }

  public String rolloutPhase() {
    return rolloutPhase;
  }

  public Integer maxPeers() {
    return maxPeers;
  }

  public Integer retryBudget() {
    return retryBudget;
  }

  public TransportPolicy transportPolicy() {
    return transportPolicy;
  }

  public AnonymityPolicy anonymityPolicy() {
    return anonymityPolicy;
  }

  public WriteModeHint writeModeHint() {
    return writeModeHint;
  }

  /**
   * Serialise the options to a JSON-compatible map. Only explicitly configured entries are
   * included.
   */
  public Map<String, Object> toJson() {
    final Map<String, Object> root = new LinkedHashMap<>();
    if (manifestEnvelopeBase64 != null) {
      root.put("manifest_envelope_b64", manifestEnvelopeBase64);
    }
    if (manifestCidHex != null) {
      root.put("manifest_cid_hex", manifestCidHex);
    }
    if (clientId != null) {
      root.put("client_id", clientId);
    }
    if (telemetryRegion != null) {
      root.put("telemetry_region", telemetryRegion);
    }
    if (rolloutPhase != null) {
      root.put("rollout_phase", rolloutPhase);
    }
    if (maxPeers != null) {
      root.put("max_peers", maxPeers);
    }
    if (retryBudget != null) {
      root.put("retry_budget", retryBudget);
    }
    if (transportPolicy != null) {
      root.put("transport_policy", transportPolicy.label());
    }
    if (anonymityPolicy != null) {
      root.put("anonymity_policy", anonymityPolicy.label());
    }
    if (writeModeHint != null && writeModeHint != WriteModeHint.READ_ONLY) {
      root.put("write_mode_hint", writeModeHint.label());
    }
    return root;
  }

  public Builder toBuilder() {
    return new Builder()
        .setManifestEnvelopeBase64(manifestEnvelopeBase64)
        .setManifestCidHex(manifestCidHex)
        .setClientId(clientId)
        .setTelemetryRegion(telemetryRegion)
        .setRolloutPhase(rolloutPhase)
        .setMaxPeers(maxPeers)
        .setRetryBudget(retryBudget)
        .setTransportPolicy(transportPolicy)
        .setAnonymityPolicy(anonymityPolicy)
        .setWriteModeHint(writeModeHint);
  }

  public static Builder builder() {
    return new Builder();
  }

  public static final class Builder {
    private String manifestEnvelopeBase64;
    private String manifestCidHex;
    private String clientId;
    private String telemetryRegion;
    private String rolloutPhase;
    private Integer maxPeers;
    private Integer retryBudget;
    // Default to SoraNet-first; callers must explicitly request DirectOnly when staging a downgrade.
    private TransportPolicy transportPolicy = TransportPolicy.SORANET_FIRST;
    private AnonymityPolicy anonymityPolicy = AnonymityPolicy.ANON_GUARD_PQ;
    private WriteModeHint writeModeHint = WriteModeHint.READ_ONLY;

    public Builder setManifestEnvelopeBase64(final String manifestEnvelopeBase64) {
      if (manifestEnvelopeBase64 == null || manifestEnvelopeBase64.trim().isEmpty()) {
        this.manifestEnvelopeBase64 = null;
      } else {
        this.manifestEnvelopeBase64 =
            SorafsInputValidator.normalizeBase64MaybeUrl(
                manifestEnvelopeBase64, "manifestEnvelopeBase64");
      }
      return this;
    }

    public Builder setManifestCidHex(final String manifestCidHex) {
      if (manifestCidHex == null || manifestCidHex.trim().isEmpty()) {
        this.manifestCidHex = null;
      } else {
        this.manifestCidHex = SorafsInputValidator.normalizeHex(manifestCidHex, "manifestCidHex");
      }
      return this;
    }

    public Builder setClientId(final String clientId) {
      this.clientId = emptyToNull(clientId);
      return this;
    }

    public Builder setTelemetryRegion(final String telemetryRegion) {
      this.telemetryRegion = emptyToNull(telemetryRegion);
      return this;
    }

    public Builder setRolloutPhase(final String rolloutPhase) {
      this.rolloutPhase = emptyToNull(rolloutPhase);
      return this;
    }

    public Builder setMaxPeers(final Integer maxPeers) {
      if (maxPeers != null && maxPeers < 1) {
        throw new IllegalArgumentException("maxPeers must be greater than zero");
      }
      this.maxPeers = maxPeers;
      return this;
    }

    public Builder setRetryBudget(final Integer retryBudget) {
      if (retryBudget != null && retryBudget < 0) {
        throw new IllegalArgumentException("retryBudget must be non-negative");
      }
      this.retryBudget = retryBudget;
      return this;
    }

    public Builder setTransportPolicy(final TransportPolicy transportPolicy) {
      this.transportPolicy = Objects.requireNonNull(transportPolicy, "transportPolicy");
      return this;
    }

    public Builder setAnonymityPolicy(final AnonymityPolicy anonymityPolicy) {
      this.anonymityPolicy = Objects.requireNonNull(anonymityPolicy, "anonymityPolicy");
      return this;
    }

    public Builder setWriteModeHint(final WriteModeHint writeModeHint) {
      this.writeModeHint = Objects.requireNonNull(writeModeHint, "writeModeHint");
      return this;
    }

    public GatewayFetchOptions build() {
      return new GatewayFetchOptions(this);
    }

    private static String emptyToNull(final String value) {
      if (value == null) {
        return null;
      }
      final String trimmed = value.trim();
      return trimmed.isEmpty() ? null : trimmed;
    }
  }
}

File: kotlin/core-jvm/src/main/java/org/hyperledger/iroha/sdk/sorafs/GatewayFetchOptions.kt
package org.hyperledger.iroha.sdk.sorafs

/**
 * Helper for composing SoraFS gateway fetch options.
 *
 * The builder exposes the subset of orchestrator configuration that Android clients commonly
 * tweak (telemetry labels, provider limits, retry budgets, transport/anonymity policy). The helper
 * serialises to a map matching the Norito JSON structure consumed by the Rust orchestrator. Use
 * `null` to omit an optional string; present strings must already be canonical and are never trimmed
 * or case-folded.
 */
class GatewayFetchOptions(
    manifestEnvelopeBase64: String? = null,
    manifestCidHex: String? = null,
    clientId: String? = null,
    telemetryRegion: String? = null,
    rolloutPhase: String? = null,
    maxPeers: Int? = null,
    retryBudget: Int? = null,
    val transportPolicy: TransportPolicy = TransportPolicy.SORANET_FIRST,
    val anonymityPolicy: AnonymityPolicy = AnonymityPolicy.ANON_GUARD_PQ,
    val writeModeHint: WriteModeHint = WriteModeHint.READ_ONLY,
) {
    val manifestEnvelopeBase64: String? = manifestEnvelopeBase64?.let {
        SorafsInputValidator.requireCanonicalBase64(it, "manifestEnvelopeBase64")
    }

    val manifestCidHex: String? = manifestCidHex?.let {
        SorafsInputValidator.requireCanonicalHexBytes(it, "manifestCidHex", 32)
    }

    val clientId: String? = exactOptional(clientId, "clientId")
    val telemetryRegion: String? = exactOptional(telemetryRegion, "telemetryRegion")
    val rolloutPhase: String? = rolloutPhase?.let {
        SorafsInputValidator.requireCanonicalRolloutPhase(it, "rolloutPhase")
    }

    val maxPeers: Int? = maxPeers?.also {
        require(it >= 1) { "maxPeers must be greater than zero" }
    }

    val retryBudget: Int? = retryBudget?.also {
        require(it >= 0) { "retryBudget must be non-negative" }
    }

    /**
     * Serialise the options to a JSON-ready map. Only explicitly configured entries are
     * included.
     */
    fun toJson(): Map<String, Any> = buildMap {
        this@GatewayFetchOptions.manifestEnvelopeBase64?.let { put("manifest_envelope_b64", it) }
        this@GatewayFetchOptions.manifestCidHex?.let { put("manifest_cid_hex", it) }
        this@GatewayFetchOptions.clientId?.let { put("client_id", it) }
        this@GatewayFetchOptions.telemetryRegion?.let { put("telemetry_region", it) }
        this@GatewayFetchOptions.rolloutPhase?.let { put("rollout_phase", it) }
        this@GatewayFetchOptions.maxPeers?.let { put("max_peers", it) }
        this@GatewayFetchOptions.retryBudget?.let { put("retry_budget", it) }
        put("transport_policy", transportPolicy.label)
        put("anonymity_policy", anonymityPolicy.label)
        if (writeModeHint != WriteModeHint.READ_ONLY) {
            put("write_mode_hint", writeModeHint.label)
        }
    }
}

private fun exactOptional(value: String?, field: String): String? =
    value?.let { SorafsInputValidator.requireExactNonEmpty(it, field) }
